File: 디크리즈컨쿼_2/selectionProblem.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#define ARR_SIZE 500
#define SWAP(a,b,t) {t=a; a=b; b=t;}
int arr[ARR_SIZE]={0};
void select(int,int,int);
int main(void){
    srand(time(NULL));
    for(int i=0;i<ARR_SIZE;i++){
        arr[i]=rand()%ARR_SIZE;
    }
    int k;
    printf("몇 번째 원소를 찾고 싶은지 입력하세요 : ");
    scanf("%d",&k);
    for(int i=0;i<ARR_SIZE;i++){
        printf("%d ",arr[i]);
    }
    select(0,ARR_SIZE-1,k);

    return 0;
}
void select(int left, int right, int k){
    if(left>=right)return;
    int pivot=arr[right];//맨 뒤의원소를 피벗으로 지정
    int low=left,high=right-1;
    int temp;
    while(low<=high){
        while(low<right && pivot>=arr[low])low++;
        while(high>=left && pivot<=arr[high])high--;
        if(low<=high){
            SWAP(arr[low],arr[high],temp);
        }
        else{
            SWAP(arr[right],arr[low],temp);
        }
    }
    if(low==k){
        printf("\n%d번째 원소는 %d\n",k,arr[k]);
    }
    else if(low<k){
        select(low+1,right,k);
    }
    else if(low>k){
        select(left,low-1,k);
    }
}

File: 디크리즈컨쿼_2/selectionProblem.py
import random
ARR_SIZE=500
arr=[]
for i in range(ARR_SIZE):
    arr.append(random.randint(0,ARR_SIZE-1))

def select(arr,left,right,k):
    if left>=right: return
    pivot=arr[right]#맨 뒤 원소가 피벗
    low=left
    high=right-1
    while low<=high:
        while low<right and pivot>=arr[low] : low+=1
        while high>=left and pivot<=arr[high] : high-=1
        if low<=high : arr[low],arr[high] = arr[high],arr[low]
        else : arr[low],arr[right] = arr[right],arr[low]
    if(k==low):
        print("\n%d번째 원소는 %d"%(k,arr[k]))
    elif(low<k):
        select(arr,low+1,right,k)
    elif(low>k):
        select(arr,left,low-1,k)
        
k=int(input("몇 번째 원소를 찾고 싶은지 입력하세요 : "))
print(arr)
select(arr,0,ARR_SIZE-1,k)